demos/coreMQTT/mqtt_demo_basic_tls.c

```diff
@@ -0,0 +1,1012 @@
+    ( void ) awsIotMqttMode;
+    ( void ) pIdentifier;
+    ( void ) pNetworkServerInfo;
+    ( void ) pNetworkCredentialInfo;
+    ( void ) pNetworkInterface;
+
+    /* Set the entry time of the demo application. This entry time will be used
+     * to calculate relative time elapsed in the execution of the demo application,
+     * by the timer utility function that is provided to the MQTT library.
+     */
+    ulGlobalEntryTimeMs = prvGetTimeMs();
+
+    for( ; ulDemoRunCount < democonfigMQTT_MAX_DEMO_COUNT; ulDemoRunCount++ )
+    {
+        /****************************** Connect. ******************************/
+
+        /* Attempt to establish a TLS connection with the MQTT broker. This example
+         * connects to the MQTT broker specified in democonfigMQTT_BROKER_ENDPOINT,
+         * using the port number specified in democonfigMQTT_BROKER_PORT (these
+         * macros are defined in file demo_config.h). If the connection fails,
+         * attempt to re-connect after a timeout. The timeout value will be
+         * exponentially increased until either the maximum timeout value is
+         * reached, or the maximum number of attempts are exhausted. The function
+         * returns a failure status if the TCP connection cannot be established
+         * with the broker after a configured number of attempts. */
+        xNetworkStatus = prvConnectToServerWithBackoffRetries( &xNetworkContext );
+
+        if( xNetworkStatus != TRANSPORT_SOCKET_STATUS_SUCCESS )
+        {
+            xDemoStatus = pdFAIL;
+        }
+        else
+        {
+            /* Set a flag indicating a TCP connection exists. This is done to
+             * disconnect if the loop exits before disconnection happens. */
+            xIsConnectionEstablished = pdTRUE;
+        }
+
+        if( xDemoStatus == pdPASS )
+        {
+            /* Send an MQTT CONNECT packet over the established TLS connection,
+             * and wait for the connection acknowledgment (CONNACK) packet. */
+            LogInfo( ( "Creating an MQTT connection to %s.\r\n", democonfigMQTT_BROKER_ENDPOINT ) );
+            xDemoStatus = prvCreateMQTTConnectionWithBroker( &xMQTTContext, &xNetworkContext );
+        }
+
+        /***************************** Subscribe. *****************************/
+
+        if( xDemoStatus == pdPASS )
+        {
+            /* If the server rejected the subscription request, attempt to resubscribe to the
+             * topic. Attempts are made according to the exponential backoff retry strategy
+             * implemented in retryUtils. */
+            xDemoStatus = prvMQTTSubscribeWithBackoffRetries( &xMQTTContext );
+        }
+
+        /******************** Publish and Keep-Alive Loop. ********************/
+
+        if( xDemoStatus == pdPASS )
+        {
+            /* Publish messages with QoS2, and send and process keep-alive messages. */
+            for( ulPublishCount = 0; ulPublishCount < ulMaxPublishCount; ulPublishCount++ )
```

Comment:
Sorry for there being so many shifts, could you change this to 
```suggestion
        for( ulPublishCount = 0; ( ulPublishCount < ulMaxPublishCount ) && ( xDemoStatus == pdPASS ); ulPublishCount++ )
```
So that we don;t keep looping when a publish fails.

Reply:
Updated, thanks.